sw/device/mask_rom/mask_rom.ld

```diff
@@ -0,0 +1,197 @@
+_test_reserved_start = _test_reserved_end - _test_reserved_size;
+
+/* Reserving space near the top of the RAM for the stack. */
+_stack_size  = 0x2000 - _test_reserved_size;
+_stack_end   = _test_reserved_start;
+_stack_start = _stack_end - _stack_size;
+
+/* Reserving 128 bytes at the top of ROM for chip info */
+_chip_info_size  = 0x80;
+_chip_info_end   = ORIGIN(rom) + LENGTH(rom);
+_chip_info_start = _chip_info_end - _chip_info_size;
+
+/**
+ * NOTE: We have to align each section to word boundaries as our current
+ * s19->slm conversion scripts are not able to handle non-word aligned sections.
+ */
+SECTIONS {
+  /**
+   * Ibex interrupt vector. See mask_rom_init.S for more information.
+   *
+   * This has to be set up at the boot address, so that execution jumps to the
+   * reset handler correctly.
+   */
+  .vectors _mask_rom_boot_address : ALIGN(256) {
+    KEEP(*(.vectors))
+    *(.vectors)
+  } > rom
+
+  /**
+   * C runtime (CRT) section, containing program initialization code.
+   *
+   * This is a separate section to `.text` so that the jumps inside `.vectors`
+   * will fit into the instruction encoding.
+   */
+  .crt : ALIGN(4) {
+    KEEP(*(.crt))
+    *(.crt)
+  } > rom
+
+  /**
+   * Standard text section, containing program code.
+   */
+  .text : ALIGN(4) {
+    *(.text)
+    *(.text.*)
+  } > rom
+
+  /**
+   * Read-only data section, containing all large compile-time constants, like
+   * strings.
+   */
+  .rodata : ALIGN(4) {
+    /* Small read-only data comes before regular read-only data for the same
+     * reasons as in the data section */
+    *(.srodata)
+    *(.srodata.*)
+    *(.rodata)
+    *(.rodata.*)
+  } > rom
+
+  /**
+   * Mutable data section, at the bottom of ram_main. This will be initialized
+   * from rom at runtime by the CRT.
+   *
+   * Load this by copying the bytes from [_data_init_start, _data_init_end] into
+   * the range [_data_start, _data_end].
+   */
+  .data ORIGIN(ram_main) : ALIGN(4) {
+    _data_start = .;
+    _data_init_start = LOADADDR(.data);
+
+    /* This will get loaded into `gp`, and the linker will use that register for
+     * accessing data within [-2048,2047] of `__global_pointer$`.
+     *
+     * This is much cheaper (for small data) than materializing the
+     * address and loading from that (which will take one extra instruction).
+     */
+    __global_pointer$ = . + 2048;
+
+    /* Small data should come before larger data. This helps to ensure small
+     * globals are within 2048 bytes of the value of `gp`, making their accesses
+     * hopefully only take one instruction. */
+    *(.sdata)
+    *(.sdata.*)
+
+    /* Other data will likely need multiple instructions to load, so we're less
+     * concerned about address materialisation taking more than one instruction.
+     */
+    *(.data)
+    *(.data.*)
+
+    /* Ensure section end is byte-aligned. */
+    . = ALIGN(4);
+    _data_end = .;
+    _data_init_end = LOADADDR(.data) + SIZEOF(.data);
```

Comment:
Would it make sense to keep `_data_init_start` and `_data_init_end` together either at the start of this section or here (for readability) since they don't depend on the location counter?

Reply:
Notionally `_data_init_end` is here, it's just we cannot use the location counter in `LOADADDR`, which is why I wanted to keep the two end symbols together.